### PR TITLE
New blocktype GtfsRTInterpreter is implemented in std-extension (UAC-1 of #219)

### PR DESCRIPTION
**Pull request by schlingling**
Introduces a new blocktype GtfsRTInterpreter.

Uses package [gtfs-realtime-binding](https://github.com/MobilityData/gtfs-realtime-bindings/tree/master) published under Apache2.0 license

Part of [RFC0006 GTFS-RT Support](https://github.com/jvalue/jayvee/tree/main/rfc/0006-gtfs-rt-support).

Implements following UACs of #219:
- [x] [UAC-1]  New blocktype GtfsRTInterpreter is implemented in std-extension.
- [x] [UAC-1.1] A new demo pipeline `gtfs-rt-simple.jv` is implemented.

This PR adds a gtfs-simple-pipeline for show casing  Example pipeline for GTFS and GTFS-RT

The pipeline works as expected, which can be seen in following Logging-Output (Note: Append Data feauture comes later).
```
[GtfsRTSimplePipeline] Overview:
        Blocks (12 blocks with 3 pipes):
         -> GTFSRTTripUpdateFeedExtractor (HttpExtractor)
                 -> GtfsRTTripUpdateInterpreter (GtfsRTInterpreter)
                         -> TripUpdateTableInterpreter (TableInterpreter)
                                 -> TripUpdateLoader (SQLiteLoader)
         -> GTFSRTVehiclePositionFeedExtractor (HttpExtractor)
                 -> GtfsRTVehiclePositionInterpreter (GtfsRTInterpreter)
                         -> VehiclePositionTableInterpreter (TableInterpreter)
                                 -> VehicleLoader (SQLiteLoader)
         -> GTFSRTAlertFeedExtractor (HttpExtractor)
                 -> GtfsRTAlertInterpreter (GtfsRTInterpreter)
                         -> AlertTableInterpreter (TableInterpreter)
                                 -> AlertLoader (SQLiteLoader)
[GTFSRTAlertFeedExtractor] Fetching raw data from https://proxy.transport.data.gouv.fr/resource/bibus-brest-gtfs-rt-alerts
[GTFSRTAlertFeedExtractor] Successfully fetched raw data
[GtfsRTAlertInterpreter] Parsing raw gtfs-rt feed data as Alerts"
[AlertTableInterpreter] Matching header with provided column names
[AlertTableInterpreter] Validating 36 row(s) according to the column types
[AlertTableInterpreter] Validation completed, the resulting table has 36 row(s) and 7 column(s)
[AlertLoader] Opening database file ./gtfs.sqlite
[AlertLoader] Dropping previous table "gtfs-rt-alert" if it exists
[AlertLoader] Creating table "gtfs-rt-alert"
[AlertLoader] Inserting 36 row(s) into table "gtfs-rt-alert"
[AlertLoader] The data was successfully loaded into the database
[GTFSRTVehiclePositionFeedExtractor] Fetching raw data from https://proxy.transport.data.gouv.fr/resource/bibus-brest-gtfs-rt-vehicle-position
[GTFSRTVehiclePositionFeedExtractor] Successfully fetched raw data
[GtfsRTVehiclePositionInterpreter] Parsing raw gtfs-rt feed data as VehiclePosition"
[VehiclePositionTableInterpreter] Matching header with provided column names
[VehiclePositionTableInterpreter] Validating 100 row(s) according to the column types
[VehiclePositionTableInterpreter] Validation completed, the resulting table has 100 row(s) and 10 column(s)
[VehicleLoader] Opening database file ./gtfs.sqlite
[VehicleLoader] Dropping previous table "gtfs-rt-vehicle_position" if it exists
[VehicleLoader] Creating table "gtfs-rt-vehicle_position"
[VehicleLoader] Inserting 100 row(s) into table "gtfs-rt-vehicle_position"
[VehicleLoader] The data was successfully loaded into the database
[GTFSRTTripUpdateFeedExtractor] Fetching raw data from https://proxy.transport.data.gouv.fr/resource/bibus-brest-gtfs-rt-trip-update
[GTFSRTTripUpdateFeedExtractor] Successfully fetched raw data
[GtfsRTTripUpdateInterpreter] Parsing raw gtfs-rt feed data as TripUpdate"
[TripUpdateTableInterpreter] Matching header with provided column names
[TripUpdateTableInterpreter] Validating 20179 row(s) according to the column types
[TripUpdateTableInterpreter] Validation completed, the resulting table has 20179 row(s) and 10 column(s)
[TripUpdateLoader] Opening database file ./gtfs.sqlite
[TripUpdateLoader] Dropping previous table "gtfs-rt-trip_update" if it exists
[TripUpdateLoader] Creating table "gtfs-rt-trip_update"
[TripUpdateLoader] Inserting 20179 row(s) into table "gtfs-rt-trip_update"
[TripUpdateLoader] The data was successfully loaded into the database
```

### CODE REVIEW

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (rhazn) [inline libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:119]: There is no need to build the object with string keys if you just remove them here again.

**Comment by rhazn** [comment]
(I assume this is a review for feedback as you wrote btw, if you want to merge this there are checks failing 😅 ).

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (schlingling) [reply libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:85]: You are right, i'll change the interface to a class with next commit.

File: libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts
Comment (schlingling) [reply libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts:33]: Added example with next commit

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (schlingling) [reply libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:91]: changed

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (schlingling) [reply libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:119]: For easier maintainment of code i decided to continue using the object with keys. Easier to handle and no need to take care of having the right index.

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (schlingling) [reply libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts]: added in commit

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (rhazn) [inline libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:114]: Which entity? Include more information like the entity type.

**Review by rhazn — CHANGES_REQUESTED** [review]
I left some comments, in general I think it is fine but the setup with the classes as attributes that then get instantiates has to go.

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (schlingling) [reply libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:114]: changed to more meaningfull errormessage.

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (rhazn) [reply libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:202]: I think in general this is the wrong way to model it. I did a bit of research and I think a record utility class might be nicer, see: https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type

Consider the `type CatName` there as a place where you can put your hardcoded list of column names. Then you can model each row of the sheet as a record of that type and string (instead of CatName and CatInfo as is done in the example). I think that might be a better solution. Using classes seemed okay for me when we talked about this because I did not really know a better way but conceptually (and as done here) it is wrong.

Also, if you do not export a class it is already private to that file. So you could just put the classes in this file and not export them so they are only available here.

**Review by rhazn — CHANGES_REQUESTED** [review]

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (schlingling) [reply libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:202]: This looks reeeally cool, got your point, thanks!

Then actually i would model it like this:
```
type TripUpdateHeader =
  | 'header.gtfs_realtime_version'
  | 'header.timestamp'
  | '...';
const row: Record<TripUpdateHeader, string> = {
  'header.gtfs_realtime_version': 'asdf',
  'header.timestamp': 'asdf',
  '...': 'asdf',
};
```

Comment (rhazn) [reply libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:202]: Yeah exactly.

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (rhazn) [inline libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:244]: What stops you from object.values if you can use entries on the record?

**Review by rhazn — APPROVED** [review]
I'll approve this because I think it is pretty self contained, now that @felix-oq is back I would love to get a second opinion from him though. So I'll request from him as well and ask you to wait with merging until he had a look. But if you do not feel the PR would impact anything we do for AMSE @felix-oq it would be nice to approve it.

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (felix-oq) [inline libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:100]: You can avoid deep nesting by negating the if conditions, e.g.
```
if (!entity.tripUpdate) {
  // log that tripUpdate was not present
  continue;
}
if (!entity.tripUpdate.stopTimeUpdate) {
  // log that stopTimeUpdate was not present
  continue;
}
// ...
```

The same technique can also be applied to your other functions.

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (rhazn) [reply libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:123]: Yes of course, sorry I thought that is what String would do if the value is undefined. Must be an empty string, not the string undefined.

File: libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts
Comment (schlingling) [reply libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts:67]: Adapted them from the docs. Do you mean each enumarated Item with a individual link or one general link to the doc? 

The element index is defined here (https://developers.google.com/transit/gtfs-realtime/reference#element-index)

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (rhazn) [inline libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts:72]: Please include the argument (`entityType`). A better error message would be `Entity ${entityType} not allowed for block GtfsRTInterpreterblock, expected "trip_update", "alert" or "vehicle".`

File: libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts
Comment (schlingling) [reply libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts]: I am not sure, what is the correct naming. Google uses CamelCase for their Elementindex (https://developers.google.com/transit/gtfs-realtime/reference#element-index) but snake_case for their field names.

**Review by rhazn — APPROVED** [review]

File: libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts
Comment (schlingling) [reply libs/extensions/std/exec/src/gtfs-rt-interpreter-executor.ts]: Thanks, changed the errormessage. As far as i understood it correct, with my new validation function the default case of this switch statement should actually never be executet. 

`l[ibs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts](https://github.com/jvalue/jayvee/blob/bc25b49c166ff0defe05be2b45cfe44d1a373830/libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts#L108)`

File: libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts
Comment (schlingling) [reply libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts:72]: changed it.

File: libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts
Comment (felix-oq) [reply libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts]: Alright, but then it should still be called `TripUpdate` without `s`.

File: libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts
Comment (felix-oq) [inline libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts:28]: To make sure that markdown recognizes the link:
```suggestion
            We currently support following Output-Sheets, each are an equivalent to the flattened Element Index defined in <https://developers.google.com/transit/gtfs-realtime/reference#element-index> (just required fields are included)):
```

File: libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts
Comment (felix-oq) [inline libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts:101]: Awesome :+1:

**Review by felix-oq — APPROVED** [review]

File: libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts
Comment (schlingling) [reply libs/extensions/std/lang/src/gtfs-rt-interpreter-meta-inf.ts]: changed

**Comment by rhazn** [comment]
I'm on my way/at a software campus event until next week. If @felix-oq is up for it/has time I'm happy if you sync with him and merge once he did a positive review but otherwise I'm back early next week.

**Comment by schlingling** [comment]
> I'm on my way/at a software campus event until next week. If @felix-oq is up for it/has time I'm happy if you sync with him and merge once he did a positive review but otherwise I'm back early next week.

Felix approved before 3h, i just commitet two last typos after the last approval from him, so i think it should be fine to merge. Enjoy the time at Softwarecampus!

**Comment by rhazn** [comment]
Ah yeah if you just do typo fixes it's fine, go for it.

**Pull request merged**
